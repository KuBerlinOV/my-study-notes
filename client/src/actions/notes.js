import { database } from '../firebase/firebase';

// import { v4 as uuidv4 } from 'uuid'
//ADD_NOTE

// export const addNote = ({ topic = '', // in this case you should imagine destructuring as you are passing some object with the same data like ' const movie = {title, year, id} '
//     description = '',           //when we are passing it to the action function as an argument and destructuring it. 
//     note = '',                 //when it would be like ' const {title, year, id} = movie ' 
//     reference = '',            //Furthermore, we provide here the default values for the object keys so
//     tag = '#',
//     status = 'in progress',     //the user could leave these options empty if he wanted to.
//     createdAt = 0
// } = {}) => {
//     return {
//         type: 'ADD_NOTE',
//         note: {
//             id: uuidv4(),
//             topic,
//             description,
//             note,
//             reference,
//             tag,
//             status,
//             createdAt
//         }
//     }
// }


export const addNote = (note, libraryId) => ({
    type: 'ADD_NOTE',
    libraryId,
    note
})

export const startAddNote = ({
    libraryId = '',
    topic = '',
    description = '',
    text = '',
    reference = '',
    tag = '#',
    status = 'in progress',
    createdAt = 0
} = {}) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        const note = {
            libraryId,
            topic,
            description,
            text,
            reference,
            tag,
            status,
            createdAt
        };
        return database.ref(`users/${uid}/notes`).push(note)
            .then((ref) => {
                //we are dispatching after pushing the note to the database 
                dispatch(addNote(
                    {
                        id: ref.key,
                        libraryId,
                        ...note
                    }))
            }).catch((e) => {
                console.log(e, 'this did not work')
            });
    };
};

//EDIT_NOTE on the local store and in database

export const updateNote = (id, updates) => ({
    type: 'UPDATE_NOTE',
    id,
    updates
})

export const startUpdateNote = (id, updates) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid
        return database.ref(`users/${uid}/notes/${id}`).update(updates).then(() => {
            dispatch(updateNote(id, updates))
        }).catch((e) => {
            console.log(e, 'the note was not updated correctly')
        })
    }
}

//updateStatus on the local store and database

export const updateStatus = (id) => ({
    type: 'UPDATE_STATUS',
    id
})

export const startUpdateStatus = (id) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid
        return database.ref(`users/${uid}/notes/${id}/status`).once('value').then(
            (snapshot) => {
                snapshot.val() === 'in progress' ? database.ref(`users/${uid}/notes/${id}`).update({ status: 'mastered' })
                    : database.ref(`users/${uid}/notes/${id}`).update({ status: 'in progress' })
            }).then(() => {
                dispatch(updateStatus(id))
            }).catch((e) => {
                console.log(e, 'status failed to update')
            })
    }
}

//REMOVE_NOTE

export const removeNote = (id) => ({
    type: 'REMOVE_NOTE',
    id
})

export const startRemoveNote = (id) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid
        return database.ref(`users/${uid}/notes/${id}`).remove().then(() => {
            dispatch(removeNote(id));
        }).catch((e) => {
            console.log(e, 'this did not work')
        })
    }
}


//SET_NOTES and startSetNotes
export const setNotes = (notes) => ({
    type: 'SET_NOTES',
    notes
});


export const startSetNotes = () => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        //the second return before the database is needed in order to make sure that in the index js file it will pass
        return database.ref(`users/${uid}/notes`).once('value').then((snapshot) => {
            const notes = [];
            snapshot.forEach((childSnapshot) => {
                notes.push({
                    id: childSnapshot.key,
                    ...childSnapshot.val()
                })
            })
            dispatch(setNotes(notes))
        }).catch((e) => {
            console.log(e, 'this did not work')
        })
    }
};