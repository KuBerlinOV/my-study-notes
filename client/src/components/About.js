import React from 'react';

const About = () => {
    return (
        <main className='about-page'>
            <h1 className='hd-lg'>About</h1>
            <p class='text'>
                MyStudyNotes is  is a personal project I created during the time of self-study to put into practice what I was learning at the FreeCodeCamp and various other platforms for learning programming.
                Even though, MyStudyNotes was created for practice it also aims to simplify a personal daily task when self-learning.
                MyStudyNotes, therefore allows users to organize their learning material into libraries
                and create notes to save the knowledge and materials accumulated during their self-study.
                It also allows change the relevance of the material written in the note by alternating its status to 'mastered'.
            </p>
        </main>
    )
}

export default About;