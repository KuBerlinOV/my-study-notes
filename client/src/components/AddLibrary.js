import React from 'react';
import { connect } from 'react-redux';
import { startAddLibrary } from '../actions/libraries';
import LibraryForm from './LibraryForm';

export const AddLibrary = (props) => {

    return (
        <div className='add-lib'>
            <h2 className='hd-md'>Create Library</h2>
            <LibraryForm
                handleSubmit={(library) => {
                    props.startAddLibrary(library);
                    props.history.push('/libraries');
                }}
            />
        </div>
    )
}

const mapDispatchToProps = (dispatch) => ({
    startAddLibrary: (library) => dispatch(startAddLibrary(library))
})

export default connect(undefined, mapDispatchToProps)(AddLibrary);