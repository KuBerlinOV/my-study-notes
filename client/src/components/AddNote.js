import React from 'react';
import { connect } from 'react-redux';
import { startAddNote } from '../actions/notes';
import NoteForm from './NoteForm';


export const AddNote = ({ libraryId, startAddNote, handleModal }) => {
    return (

        <NoteForm
            handleModal={handleModal}
            libraryId={libraryId}
            handleSubmit={(note) => {
                startAddNote(note, libraryId);
            }}
        />
    )
}

const mapDispatchToProps = (dispatch) => ({
    startAddNote: (note, libraryId) => dispatch(startAddNote(note, libraryId))
})

export default connect(undefined, mapDispatchToProps)(AddNote);