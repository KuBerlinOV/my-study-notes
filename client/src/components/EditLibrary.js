import React from 'react';
import { connect } from 'react-redux';
import LibraryForm from './LibraryForm';
import { startUpdateLibrary } from '../actions/libraries';


export const EditLibrary = (props) => {
    return (
        <div>
            <LibraryForm
                //library before update
                library={props.library}
                //this is handle submit props
                handleSubmit={(library) => {
                    //dispatch action to edit the expense
                    props.startUpdateLibrary(props.library.id, library);
                    //redirect to dashBoard
                    props.history.push('/libraries')
                }}
            />
        </div>
    )
}

const mapDispatchToProps = (dispatch) => ({
    startUpdateLibrary: (id, library) => dispatch(startUpdateLibrary(id, library))
})

const mapStateToProps = (state, props) => {
    return {
        library: state.libraries.find((library) => library.id == props.match.params.id)
    }

}

export default connect(mapStateToProps, mapDispatchToProps)(EditLibrary);