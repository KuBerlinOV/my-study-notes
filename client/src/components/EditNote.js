import React from 'react';
import { connect } from 'react-redux';
import NoteForm from './NoteForm';
import { startUpdateNote } from '../actions/notes';


export const EditNote = (props) => {
    return (
        <main className='editnote-page'>
            <NoteForm
                //note before update
                note={props.note}
                //this is handle submit props
                handleSubmit={(note) => {
                    //dispatch action to edit the expense
                    props.startUpdateNote(props.note.id, note);
                    //redirect to dashBoard
                    props.history.push('/notes')
                }}
            />
        </main>
    )
}

const mapDispatchToProps = (dispatch) => ({
    startUpdateNote: (id, note) => dispatch(startUpdateNote(id, note))
})

const mapStateToProps = (state, props) => {
    return {
        note: state.notes.find((note) => note.id == props.match.params.id)
    }

}

export default connect(mapStateToProps, mapDispatchToProps)(EditNote);