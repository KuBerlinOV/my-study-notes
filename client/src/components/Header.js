import React from 'react';
import { NavLink } from 'react-router-dom';
import { startLogout } from '../actions/auth';
import { connect } from 'react-redux';


export const Header = ({ startLogout, isAuthenticated }) => (
    isAuthenticated ? (
        <div className='hd'>
            <header id='hd-private'>
                <NavLink activeClassName='is-active' className='nav-link' to='/home' exact={true}>Home</NavLink>
                <NavLink activeClassName='is-active' className='nav-link' to='/libraries'>My Libraries</NavLink>
                <NavLink activeClassName='is-active' className='nav-link' to='/notes'>All Notes</NavLink>
                <NavLink activeClassName='is-active' className='nav-link' to='/about'>About</NavLink>
                <button className='logout-btn' onClick={startLogout} >Logout</button>
            </header>
        </div>
    ) : (
        <div className='hd'>
            <header id='hd-public'>
                <NavLink activeClassName='is-active' className='nav-link' to='/' exact={true}>Login</NavLink>
                <NavLink activeClassName='is-active' className='nav-link' to='/about'>About</NavLink>
            </header>
        </div>
    )
);

const mapStateToProps = (state) => ({
    isAuthenticated: !!state.auth.uid
})

const mapDispatchToProps = (dispatch) => ({
    startLogout: () => { dispatch(startLogout()) }
})

export default connect(mapStateToProps, mapDispatchToProps)(Header);
