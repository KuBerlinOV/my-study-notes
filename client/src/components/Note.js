import React, { useState } from 'react';
import { connect } from 'react-redux';
import moment from 'moment';
import { startRemoveNote, startUpdateStatus } from '../actions/notes'
import Modal from 'react-modal';
//here I imported the history object created in approuter, as note does not have access to it, as it is not rendered in the approuter. It is out of the context
import { history } from '../routers/approuter';
// note: {
//     id: uuid(),
//     topic,
//     description,
//     note,
//     reference,
//     tag,
//     status,
//     createdAt
// }

export const Note = ({ topic, description, text, reference, tag, createdAt, id, status, startUpdateStatus, startRemoveNote }) => {
    // const history = useHistory(); // use this instead of link to redirect to another page with params
    const [openModal, setOpenModal] = useState(false)

    const handleModal = (e) => {
        e.preventDefault();
        setOpenModal(!openModal)
    }


    return (
        <div className='note-card'>
            <h2 className='hd-md'> {topic} </h2>
            <hr />
            <div className='note-card-info'>
                <p className='info-text' id='note-text'>{text} </p>
                <hr />
                <p className='info-text'>Reference: {reference} </p>
                <p className='info-text'>Tag: {tag} </p>
                <p className='info-text'>Status: {status}</p>
                <p className='info-text'>{moment(createdAt).format('MMMM Do YYYY, h:mm')} </p>
            </div>
            <hr />
            <div className='note-card-btn'>
                <button onClick={() => {
                    startUpdateStatus(id)
                }}>Change status</button>
                <button onClick={handleModal}>Delete</button>
                <button onClick={() => {
                    history.push(`/editnote/${id}`)
                }}> Edit</button>
            </div>
            <Modal
                isOpen={openModal}
                ariaHideApp={false}
                onRequestClose={handleModal}
                className="delete-warning"
            >
                <h3>Are you sure?</h3>
                <p>Do you really want to delete this note? This process cannot be undone!</p>

                <button onClick={() => { startRemoveNote(id) }}>Delete</button>
            </Modal>
        </div >
    )
}

// <Link to={`/edit/${id}`}> Edit</Link>

const mapDispatchToProps = (dispatch) => ({
    startRemoveNote: (id) => dispatch(startRemoveNote(id)),
    startUpdateStatus: (id) => dispatch(startUpdateStatus(id))
})

export default connect(undefined, mapDispatchToProps)(Note);