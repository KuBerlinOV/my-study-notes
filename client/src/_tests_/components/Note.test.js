import React from 'react';
import { mount } from 'enzyme';
import { Note } from '../../components/Note';
import notes from '../fixtures/notes'
import Modal from 'react-modal';
//testing react modal => https://remarkablemark.org/blog/2017/05/17/testing-react-modal/

let wrapper, handleModalSpy, updateStatusSpy;

beforeEach(() => {
    handleModalSpy = jest.fn();
    updateStatusSpy = jest.fn();
    wrapper = mount(<Note updateStatus={updateStatusSpy} {...notes[0]} />);
})

describe('Note', () => {
    it('should render Note', () => {
        // const wrapper = mount(<Note {...notes[0]}
        // // topic={notes[0].topic}
        // // description={notes[0].description}
        // // note={notes[0].note}
        // // reference={notes[0].reference}
        // // tag={notes[0].tag}
        // // createdAt={notes[0].createdAt}
        // // id={notes[0].id}
        // // status={notes[0].status}
        // />)
        expect(wrapper).toMatchSnapshot();
    })
    it('renders react-modal', () => {
        expect(wrapper.find(Modal)).toHaveLength(1);
        expect(wrapper.find(Modal).prop('isOpen')).toBe(false);
    });
    it('opens modal', () => {
        wrapper.find('button').at(1).simulate('click')
        expect(wrapper.find(Modal).prop('isOpen')).toBe(true);
    })
    it('should call update note with correct id', () => {
        wrapper.find('button').at(0).simulate('click')
        expect(updateStatusSpy).toHaveBeenCalledWith(notes[0].id)
    })
})