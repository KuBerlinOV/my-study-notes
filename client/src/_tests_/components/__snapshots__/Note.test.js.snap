// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Note should render Note 1`] = `
<Note
  createdAt={0}
  description="redux note"
  id="1"
  reference="some reference"
  status="mastered"
  tag="some tag"
  text="here is a note"
  topic="redux"
  updateStatus={[MockFunction]}
>
  <div>
    <h3>
       Topic: 
      redux
       
    </h3>
    <p>
      Description: 
      redux note
       
    </p>
    <p>
      Note: 
       
    </p>
    <p>
      Reference: 
      some reference
       
    </p>
    <p>
      Tag: 
      some tag
       
    </p>
    <p>
      Status: 
      mastered
    </p>
    <button
      onClick={[Function]}
    >
      Change status
    </button>
    <p>
      Date: 
      January 1st 1970, 1:00
       
    </p>
    <button
      onClick={[Function]}
    >
      Delete
    </button>
    <button
      onClick={[Function]}
    >
       Edit
    </button>
    <Modal
      ariaHideApp={true}
      bodyOpenClassName="ReactModal__Body--open"
      className="delete-warning"
      closeTimeoutMS={0}
      contentElement={[Function]}
      isOpen={false}
      onRequestClose={[Function]}
      overlayElement={[Function]}
      parentSelector={[Function]}
      portalClassName="ReactModalPortal"
      preventScroll={false}
      role="dialog"
      shouldCloseOnEsc={true}
      shouldCloseOnOverlayClick={true}
      shouldFocusAfterRender={true}
      shouldReturnFocusAfterClose={true}
    >
      <Portal
        containerInfo={
          <div
            class="ReactModalPortal"
          />
        }
      >
        <ModalPortal
          ariaHideApp={true}
          bodyOpenClassName="ReactModal__Body--open"
          className="delete-warning"
          closeTimeoutMS={0}
          contentElement={[Function]}
          defaultStyles={
            Object {
              "content": Object {
                "WebkitOverflowScrolling": "touch",
                "background": "#fff",
                "border": "1px solid #ccc",
                "borderRadius": "4px",
                "bottom": "40px",
                "left": "40px",
                "outline": "none",
                "overflow": "auto",
                "padding": "20px",
                "position": "absolute",
                "right": "40px",
                "top": "40px",
              },
              "overlay": Object {
                "backgroundColor": "rgba(255, 255, 255, 0.75)",
                "bottom": 0,
                "left": 0,
                "position": "fixed",
                "right": 0,
                "top": 0,
              },
            }
          }
          isOpen={false}
          onRequestClose={[Function]}
          overlayElement={[Function]}
          parentSelector={[Function]}
          portalClassName="ReactModalPortal"
          preventScroll={false}
          role="dialog"
          shouldCloseOnEsc={true}
          shouldCloseOnOverlayClick={true}
          shouldFocusAfterRender={true}
          shouldReturnFocusAfterClose={true}
          style={
            Object {
              "content": Object {},
              "overlay": Object {},
            }
          }
        />
      </Portal>
    </Modal>
  </div>
</Note>
`;
